### ISSUE TITLE
Error in CUDAExecutionProvider of onnxruntime-gpu leads to unwanted modification of python environment

### ISSUE THREAD

**Issue by jens-siebert** [ultralytics/ultralytics#21092]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Other

### Bug

(might be related to #5093)

Hey there,

we're using Ultralytics to run a YOLOv11 model converted to ONNX on our production systems to detect and track palettes. Ultralytics is installed within a python environment where onnxruntime-gpu is already installed. The systems are equipped with a NVIDIA RTX A2000 GPU with 12 GB of VRAM.

On one of our systems we've encountered an issue with the GPU falling off the bus due to overheat. This resulted in an error in the CUDAExecutionProider of onnxruntime(-gpu). 

```shell
Jun 17 14:58:05 doks-gate-6 python3[382649]: *************** EP Error ***************
Jun 17 14:58:05 doks-gate-6 python3[382649]: EP Error /onnxruntime_src/onnxruntime/core/providers/cuda/cuda_call.cc:129 std::conditional_t<THRW, void, onnxruntime::common::Status> onnxruntime::CudaCall(ERRTYPE, const char*, const char*, SUCCTYPE, const char*, const char*, int) [with ERRTYPE = cudaError; bool THRW = true; SUCCTYPE = cudaError; std::conditional_t<THRW, void, common::Status> = void] /onnxruntime_src/onnxruntime/core/providers/cuda/cuda_call.cc:121 std::conditional_t<THRW, void, onnxruntime::common::Status> onnxruntime::CudaCall(ERRTYPE, const char*, const char*, SUCCTYPE, const char*, const char*, int) [with ERRTYPE = cudaError; bool THRW = true; SUCCTYPE = cudaError; std::conditional_t<THRW, void, common::Status> = void] CUDA failure 999: unknown error ; GPU=-1 ; hostname=doks-gate-6 ; file=/onnxruntime_src/onnxruntime/core/providers/cuda/cuda_execution_provider.cc ; line=282 ; expr=cudaSetDevice(info_.device_id);
Jun 17 14:58:05 doks-gate-6 python3[382649]:  when using ['CUDAExecutionProvider', 'CPUExecutionProvider']
Jun 17 14:58:05 doks-gate-6 python3[382649]: Falling back to ['CPUExecutionProvider'] and retrying.
Jun 17 14:58:05 doks-gate-6 python3[382649]: ****************************************
Jun 17 14:59:35 doks-gate-6 python3[382804]: /home/doks/repos/gate_controller/venv/lib/python3.10/site-packages/torch/cuda/__init__.py:174: UserWarning: CUDA initialization: CUDA unknown error - this may be due to an incorrectly set up environment, e.g. changing env variable CUDA_VISIBLE_DEVICES after program start. Setting the available devices to be zero. (Triggered internally at /pytorch/c10/cuda/CUDAFunctions.cpp:109.)
Jun 17 14:59:35 doks-gate-6 python3[382804]:   return torch._C._cuda_getDeviceCount() > 0
Jun 17 14:59:35 doks-gate-6 python3[382804]: Loading /home/doks/repos/gate_controller/imagedata/../data/palette_detector.onnx for ONNX Runtime inference...
```

This error led to the automatic installation of onnxruntime (non-gpu) via AutoUpdate:
```shell
Jun 17 14:59:35 doks-gate-6 python3[382804]: requirements: Ultralytics requirement ['onnxruntime'] not found, attempting AutoUpdate...
Jun 17 14:59:38 doks-gate-6 python3[383241]: [notice] A new release of pip is available: 24.3.1 -> 25.1.1
Jun 17 14:59:38 doks-gate-6 python3[383241]: [notice] To update, run: pip install --upgrade pip
Jun 17 14:59:39 doks-gate-6 python3[382804]: Collecting onnxruntime
Jun 17 14:59:39 doks-gate-6 python3[382804]:   Downloading onnxruntime-1.22.0-cp310-cp310-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl.metadata (4.5 kB)
Jun 17 14:59:39 doks-gate-6 python3[382804]: Requirement already satisfied: coloredlogs in ./venv/lib/python3.10/site-packages (from onnxruntime) (15.0.1)
Jun 17 14:59:39 doks-gate-6 python3[382804]: Requirement already satisfied: flatbuffers in ./venv/lib/python3.10/site-packages (from onnxruntime) (25.2.10)
Jun 17 14:59:39 doks-gate-6 python3[382804]: Requirement already satisfied: numpy>=1.21.6 in ./venv/lib/python3.10/site-packages (from onnxruntime) (2.1.0)
Jun 17 14:59:39 doks-gate-6 python3[382804]: Requirement already satisfied: packaging in ./venv/lib/python3.10/site-packages (from onnxruntime) (25.0)
Jun 17 14:59:39 doks-gate-6 python3[382804]: Requirement already satisfied: protobuf in ./venv/lib/python3.10/site-packages (from onnxruntime) (6.31.1)
Jun 17 14:59:39 doks-gate-6 python3[382804]: Requirement already satisfied: sympy in ./venv/lib/python3.10/site-packages (from onnxruntime) (1.14.0)
Jun 17 14:59:39 doks-gate-6 python3[382804]: Requirement already satisfied: humanfriendly>=9.1 in ./venv/lib/python3.10/site-packages (from coloredlogs->onnxruntime) (10.0)
Jun 17 14:59:39 doks-gate-6 python3[382804]: Requirement already satisfied: mpmath<1.4,>=1.1.0 in ./venv/lib/python3.10/site-packages (from sympy->onnxruntime) (1.3.0)
Jun 17 14:59:39 doks-gate-6 python3[382804]: Downloading onnxruntime-1.22.0-cp310-cp310-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl (16.4 MB)
Jun 17 14:59:39 doks-gate-6 python3[382804]:    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.4/16.4 MB 11.0 MB/s eta 0:00:00
Jun 17 14:59:39 doks-gate-6 python3[382804]: Installing collected packages: onnxruntime
Jun 17 14:59:39 doks-gate-6 python3[382804]: Successfully installed onnxruntime-1.22.0
Jun 17 14:59:39 doks-gate-6 python3[382804]: requirements: AutoUpdate success ✅ 3.7s
Jun 17 14:59:39 doks-gate-6 python3[382804]: WARNING ⚠️ requirements: Restart runtime or rerun command for updates to take effect
Jun 17 14:59:39 doks-gate-6 python3[382804]: Using ONNX Runtime CPUExecutionProvider
```

After that onnxruntime was using the CPUExecutionProvider but the python environment was altered also in a way that it made it impossible to use the CUDAExecutionProvider again after fixing the GPU overheat problem and restarting our software as the provider is no longer installed:

```shell
Jun 17 15:04:30 doks-gate-6 systemd[1]: Started summAIRy.gate controller.
Jun 17 15:04:37 doks-gate-6 python3[2029]: /home/doks/repos/gate_controller/venv/lib/python3.10/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py:121: UserWarning: Specified provider 'CUDAExecutionProvider' is not in available provider names.Available providers: 'AzureExecutionProvider, CPUExecutionProvider'
Jun 17 15:04:37 doks-gate-6 python3[2029]:   warnings.warn(
Jun 17 15:06:09 doks-gate-6 python3[2971]: Loading /home/doks/repos/gate_controller/imagedata/../data/palette_detector.onnx for ONNX Runtime inference...
Jun 17 15:06:09 doks-gate-6 python3[2971]: WARNING ⚠️ Failed to start ONNX Runtime with CUDA. Using CPU...
Jun 17 15:06:09 doks-gate-6 python3[2971]: Using ONNX Runtime CPUExecutionProvider
```

In my opinion the installation of onnxruntime (non-gpu) is completely unnecessary in this case as onnxruntime-gpu already contains the CPUExecutionProvider that it can use as a fallback. According to the log output onnxruntime-gpu would've even used this fallback but a subsequent check for CUDA availability using PyTorch seemed to have failed which in turn seemed to have led to the installation of onnxruntime (non-gpu).

### Environment

Ultralytics 8.3.151 🚀 Python-3.10.12 torch-2.7.1+cu126 CUDA:0 (NVIDIA RTX A2000 12GB, 11907MiB)
Setup complete ✅ (32 CPUs, 62.5 GB RAM, 408.5/1813.1 GB disk)

OS                  Linux-5.15.0-135-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.10.12
Install             git
Path                /home/doks/repos/gate_controller/venv/lib/python3.10/site-packages/ultralytics
RAM                 62.51 GB
Disk                408.5/1813.1 GB
CPU                 13th Gen Intel Core(TM) i9-13900E
CPU count           32
GPU                 NVIDIA RTX A2000 12GB, 11907MiB
GPU count           1
CUDA                12.6

numpy               ✅ 2.1.0>=1.23.0
matplotlib          ✅ 3.10.3>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.1.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.4>=2.23.0
scipy               ✅ 1.15.3>=1.4.1
torch               ✅ 2.7.1>=1.8.0
torch               ✅ 2.7.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.22.1>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.3.0>=1.1.4
ultralytics-thop    ✅ 2.0.14>=2.0.0

### Minimal Reproducible Example

(not possible to provide an example as this involves a hardware error)

### Additional

I do have a related question: is it possible to deactivate AutoUpdate completely or have it print missing dependencies instead of automagically installing them?

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jens-siebert, thank you for reporting this detailed issue and for your interest in contributing to Ultralytics 🚀! This is an automated response to help get things started—an Ultralytics engineer will review your report and provide further assistance soon.

We recommend checking out the [Docs](https://docs.ultralytics.com/) for general guidance, including many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, as well as answers to common questions.

Since this is a 🐛 Bug Report, if you discover a way to create a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE), even a simplified code snippet or setup, please share it here to help us investigate more efficiently.

For custom training or environment-specific issues, please include as much context as possible, such as configuration files or logs, and consider reviewing our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community to connect with other users and contributors:
- For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth technical discussions, visit [Discourse](https://community.ultralytics.com/)
- Or check out threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share and learn

## Upgrade

Please ensure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to confirm your issue is not already resolved in the latest release:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in these up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your report and willingness to contribute!

**Comment by Y-T-G** [comment]
Can you open a PR for this?

**Comment by ghost** [comment]
Yeah sure, I can have a look... 👍

**Comment by Laughing-q** [comment]
> I do have a related question: is it possible to deactivate AutoUpdate completely or have it print missing dependencies instead of automagically installing them?

@jens-siebert good point! yeah we actually have a python env for this, that you can disable all the autoinstall behavior by setting `YOLO_AUTOINSTALL=False`
Reference: https://github.com/ultralytics/ultralytics/blob/a7910a9a077a5b052df60877b214ed3462884bcb/ultralytics/utils/__init__.py#L42-L42

```bash
# this would autoinstall onnxruntime package
yolo export format=onnx model=yolo11n.pt
yolo predict model=yolo11n.onnx

# with YOLO_AUTOINSTALL=False it'd not install onnxruntime
YOLO_AUTOINSTALL=False yolo predict model=yolo11n.onnx
```

**Comment by ghost** [comment]
Thanks for the hint @Laughing-q 👍

**Comment by Laughing-q** [comment]
@jens-siebert Hope this resolve the issue for you! I'm going to close the issue and probably that PR as well. Please feel free to reopen the issue/PR if you feel so. Thanks again for your contribution on this.
